src/winsvc.c

```diff
@@ -0,0 +1,633 @@
+  baton->next = NULL;
+  return baton;
+}
+
+static void svc_destroy_baton(lua_State* L, svc_baton* baton) {
+  luaL_unref(L, LUA_REGISTRYINDEX, baton->block.lua_cb_ref);
+  luaL_unref(L, LUA_REGISTRYINDEX, baton->lua_main_ref);
+  free(baton->name);
+  uv_close((uv_handle_t*)&baton->svc_async_handle, NULL);
+  uv_close((uv_handle_t*)&baton->block.async_handle, NULL);
+  CloseHandle(baton->svc_end_event);
+  CloseHandle(baton->block.block_end_event);
+  LocalFree(baton);
+}
+
+static svc_baton* find_baton(const char* name) {
+  svc_baton* it = gBatons;
+  while (it != NULL) {
+    if (strcmp(it->name, name) == 0) {
+      break;
+    }
+    it = it->next;
+  }
+
+  return it;
+}
+
+
+VOID WINAPI ServiceMain(_In_  DWORD dwArgc, _In_  LPTSTR *lpszArgv) {
+  svc_baton *baton = find_baton(lpszArgv[0]);
+  baton->status_handle = RegisterServiceCtrlHandlerEx(baton->name, HandlerEx, baton);
+  baton->dwArgc = dwArgc;
+  baton->lpszArgv = lpszArgv;
+
+  uv_async_send(&baton->svc_async_handle);
+  WaitForSingleObject(baton->svc_end_event, INFINITE);
+}
+
+static int lua_GetStatusHandleFromContext(lua_State *L) {
+  svc_baton* baton = lua_touserdata(L, 1);
+  lua_pushlightuserdata(L, baton->status_handle);
+  return 1;
+}
+
+static int lua_EndService(lua_State *L) {
+  svc_baton* baton = lua_touserdata(L, 1);
+  SetEvent(baton->svc_end_event);
+  return 0;
+}
+
+static int lua_SetServiceStatus(lua_State *L) {
+  SERVICE_STATUS status;
+  SERVICE_STATUS_HANDLE SvcCtrlHandler = lua_touserdata(L, 1);
+  if (!lua_istable(L, 2)) {
+    return luaL_error(L, "table expected");
+  }
+
+  status.dwCheckPoint = GetDWFromTable(L, "dwCheckPoint");
+  status.dwControlsAccepted = GetDWFromTable(L, "dwControlsAccepted");
+  status.dwCurrentState = GetDWFromTable(L, "dwCurrentState");
+  status.dwServiceSpecificExitCode = GetDWFromTable(L, "dwServiceSpecificExitCode");
+  status.dwServiceType = GetDWFromTable(L, "dwServiceType");
+  status.dwWaitHint = GetDWFromTable(L, "dwWaitHint");
+  status.dwWin32ExitCode = GetDWFromTable(L, "dwWin32ExitCode");
+
+  BOOL ret = SetServiceStatus(SvcCtrlHandler, (LPSERVICE_STATUS)&status);
+  if (ret) {
+    lua_pushboolean(L, ret);
+    lua_pushnil(L);
+  }
+  else {
+    lua_pushnil(L);
+    lua_pushinteger(L, GetLastError());
+  }
+  return 2;
+}
+
+static void svcdispatcher_end_cb(uv_handle_t* handle) {
+  svc_dispatch_info *info = (svc_dispatch_info*)handle->data;
+  lua_State* L = luv_state(handle->loop);
+
+  /* Cleanup baton linked list */
+  svc_baton *svc_baton_it = gBatons;
+  while (svc_baton_it != NULL) {
+    svc_baton *old = svc_baton_it;
+    svc_baton_it = svc_baton_it->next;
+    svc_destroy_baton(L, old);
+  }
+
+  uv_close((uv_handle_t*)&info->end_async_handle, NULL);
+  gBatons = NULL;
+
+  lua_rawgeti(L, LUA_REGISTRYINDEX, info->lua_cb_ref);
+  lua_pushboolean(L, info->return_code);
+  if (info->return_code) {
+    lua_pushnil(L);
+  }
+  else {
+    lua_pushinteger(L, info->error);
+  }
+  lua_call(L, 2, 0);
```

Comment:
If there is an error in the lua code, then the C code could display it in a nice manner or take some other action, instead of just crashing and leaving the service in an unknown state